aws/bucket.ts

```diff
@@ -80,15 +84,18 @@ export class Bucket extends pulumi.ComponentResource implements cloud.Bucket {
 
     }
 
-    public onPut(name: string, handler: cloud.BucketHandler, filter?: cloud.BucketFilter) {
+    public onPut(name: string, handler: BucketHandler, filter?: cloud.BucketFilter) {
         this.addHandler(name, handler, ["s3:ObjectCreated:*"], filter);
     }
 
-    public onDelete(name: string, handler: cloud.BucketHandler, filter?: cloud.BucketFilter) {
+    public onDelete(name: string, handler: BucketHandler, filter?: cloud.BucketFilter) {
         this.addHandler(name, handler, ["s3:ObjectRemoved:*"], filter);
     }
 
-    public addHandler(name: string, handler: cloud.BucketHandler, events: string[], filter?: cloud.BucketFilter) {
+    public addHandler(name: string, handler: BucketHandler, events: string[], filter?: cloud.BucketFilter) {
+        const data = callback.getOrCreateAwsCallbackData(handler);
+        const handlerFunc = data.function;
```

Comment:
the pattern in all classes is to take in the AwsCallback/AzureCallback, and then create the 'data' object that describes the Lambda/FunctionApp to create.  If the value is a simple Javascript Function instance, then this will return a data object pointing at that instance, with the pulumi default values for everything else.  If it was *already* a real 'data' object, then that is returned untouched.